lib/fileset/default.nix

```diff
@@ -0,0 +1,175 @@
+# Note: Derive property tests from https://en.wikipedia.org/wiki/Algebra_of_sets
+{ lib }: {
+  /*
+  Create a file set from a path:
+  - If the path points to a file, the result is a file set only consisting of that single file
+  - If the path points to a directory, all files in that directory will be in the resulting file set
+
+  This function is effectively not needed because all functions in `lib.fileset` that expect a file set will implicitly apply the `singleton` function when a path is passed instead.
+
+  Type:
+    singleton :: Path -> FileSet
+  */
+  singleton = null;
+
+  /*
+  Trace all files included in the file set in a pretty way. Note that this strictly evaluates the file set. Example output:
+
+      trace: /home/user/src/myProject
+      trace: - Makefile
+      trace: - src
+      trace:   - main.c
+
+  Type:
+    trace :: FileSet -> Any -> Any
+  */
+  trace = null;
+
+  /*
+  Create a file set containing all files tracked by a local git repository
+
+  Type:
+    gitTracked :: Path -> FileSet
+  */
+  gitTracked = null;
+
+  /*
+  Limit a file set to only the files with specific extensions.
+
+      fileExtensionFilter extensionList
+
+  is equivalent to
+
+      fileFilter ({ extension, ... }: lib.elem extension extensionList)
+
+  Type:
+    fileExtensionFilter :: [String] -> FileSet -> FileSet
+  */
+  fileExtensionFilter = null;
+
+  /*
+  Limit a file set to only the files satisfying a predicate function.
+
+  Type:
+    fileFilter :: ({
+      path :: Path,
+      root :: Path,
+      components :: [String],
+      name :: String,
+      type :: String,
+      extension :: Null || String,
+      ...
+    } -> Bool) -> FileSet -> FileSet
```

Comment:
Which attributes should be exposed here? I don't think there's any harm to having all of these, it makes writing file filters easier, though I don't think they will be necessary often.